src/ast/passes/simplify_types.cpp

```diff
@@ -91,11 +92,49 @@ std::optional<Expression> SimplifyTypes::visit(Binop &op)
   // and alignment issus.
   // https://github.com/bpftrace/bpftrace/pull/4523#discussion_r2382109017
   std::vector<Expression> equal_exprs;
+  StatementList block_stmts;
+
+  Expression left_tuple;
+  Expression right_tuple;
+
+  // Assign tuples that aren't variables or map accesses to a temp variable
+  // so that the entire tuple can be evaluated before comparison --
+  // consider the case of `$a = 1; ({ 1 + $a }, 3) == (2, 3)`; these should
+  // be equal
+  var_id_++;
+  if (op.left.is<Variable>() || op.left.is<MapAccess>()) {
+    left_tuple = clone(ast_, op.left, op.left.loc());
+  } else {
+    left_tuple = ast_.make_node<Variable>("$$binop_tuple_left_" +
+                                              std::to_string(var_id_),
+                                          Location(op.left.loc()));
+    auto *left_var_assign = ast_.make_node<AssignVarStatement>(
+        left_tuple.as<Variable>(),
```

Comment:
You need to clone this also. (Here and below.)

Reply:
good catch. done.